runtime/parachains/src/paras_inherent.rs

```diff
@@ -290,239 +459,1658 @@ pub mod pallet {
+		})
+	}
+}
+
+/// Assures the `$condition` is `true`, and raises
+/// an error if `$action` is `true`.
+/// If `$action` is `false`, executes `$alt` if present.
+macro_rules! ensure2 {
+	($condition:expr, $err:expr, $action:ident $(, $alt:expr)? $(,)?) => {
+		let condition = $condition;
+		if !condition {
+			if $action {
+				ensure!(condition, $err);
+			} else {
+				$($alt)?
+			}
+		}
+	};
+}
+
+/// Derive a bitfield from dispute
+pub(super) fn create_disputed_bitfield<'a, I>(
+	expected_bits: usize,
+	freed_cores: I,
+) -> DisputedBitfield
+where
+	I: 'a + IntoIterator<Item = &'a CoreIndex>,
+{
+	let mut bitvec = BitVec::repeat(false, expected_bits);
+	for core_idx in freed_cores {
+		let core_idx = core_idx.0 as usize;
+		if core_idx < expected_bits {
+			bitvec.set(core_idx, true);
+		}
+	}
+	DisputedBitfield::from(bitvec)
+}
+
+fn random_sel<X, F: Fn(&X) -> Weight>(
+	rng: &mut rand_chacha::ChaChaRng,
+	selectables: Vec<X>,
+	weight_fn: F,
+	weight_limit: Weight,
+) -> (Weight, Vec<usize>) {
+	if selectables.is_empty() {
+		return (0 as Weight, Vec::new())
+	}
+	let mut indices = (0..selectables.len()).into_iter().collect::<Vec<_>>();
+	let mut picked_indices = Vec::with_capacity(selectables.len().saturating_sub(1));
+
+	let mut weight_acc = 0 as Weight;
+	while !indices.is_empty() {
+		// randomly pick an index
+		let pick = rng.gen_range(0..indices.len());
+		// remove the index from the available set of indices
+		let idx = indices.swap_remove(pick);
+
+		let item = &selectables[idx];
+		weight_acc += weight_fn(item);
+
+		if weight_acc > weight_limit {
+			break
+		}
+
+		picked_indices.push(idx);
+	}
+
+	// sorting indices, so the ordering is retained
+	// unstable sorting is fine, since there are no duplicates
+	picked_indices.sort_unstable();
+	(weight_acc, picked_indices)
+}
+
+/// Considers an upper threshold that the candidates must not exceed.
 ///
-/// Use a configured assumption about the weight required to process a backed candidate and the
-/// current block weight as of the execution of this function to ensure that we don't overload
-/// the block with candidate processing.
+/// If there is sufficient space, all bitfields and candidates will be included.
 ///
-/// If the backed candidates exceed the available block weight remaining, then skips all of them.
-/// This is somewhat less desirable than attempting to fit some of them, but is more fair in the
-/// even that we can't trust the provisioner to provide a fair / random ordering of candidates.
-fn limit_backed_candidates<T: Config>(
-	mut backed_candidates: Vec<BackedCandidate<T::Hash>>,
-) -> Vec<BackedCandidate<T::Hash>> {
-	const MAX_CODE_UPGRADES: usize = 1;
+/// Otherwise tries to include all bitfields, and fills in the remaining weight with candidates.
+///
+/// If even the bitfields are too large to fit into the `max_weight` limit, bitfields are randomly
+/// picked and _no_ candidates will be included.
+fn apply_weight_limit<T: Config + inclusion::Config>(
```

Comment:
@Lldenaurois we should probably move the disputes weight limiting in here as well, to make it an all-in-one weight limiting, so there cannot be ordering problems. It also de-duplicates some of the `RNG` creation boilerplate.

Reply:
if do this we should have this function call other functions like `limit_bitfields`, `limit_candidates`, and `limit_disputes`, so each can be unit tested.

Reply:
Indeed, very good point